File: packages/sdk-py/promptops/prompt.py
# ─────────────────────────────────────────────
#  PromptOps Python SDK — Prompt
#  The object returned from client.get()
# ─────────────────────────────────────────────

from __future__ import annotations
import re
from typing import Any, Callable, Awaitable
from .types import RawPrompt, PromptMessage, PromptMetadata, TestCase, TestResult
from .exceptions import PromptOpsCompileError


class Prompt:
    """
    Represents a fetched prompt with its metadata and messages.
    Use .compile() to interpolate variables and get messages ready for any LLM.
    """

    def __init__(self, raw: RawPrompt):
        self.metadata: PromptMetadata = raw.metadata
        self._raw_messages = raw.messages

    def compile(
        self,
        variables: dict[str, Any] | None = None,
        trim: bool = True,
    ) -> list[dict[str, str]]:
        """
        Interpolate {{variable}} placeholders and return messages
        ready for any LLM provider.

        Example:
            messages = prompt.compile(email=email_body, tone="formal")
            response = openai.chat.completions.create(
                model="gpt-4o",
                messages=messages,
            )
        """
        variables = variables or {}

        # Check for missing required variables
        missing = [v for v in self.metadata.variables if v not in variables]
        if missing:
            raise PromptOpsCompileError(
                f'Missing required variables for prompt "{self.metadata.slug}": '
                f'{", ".join(missing)}'
            )

        return [
            {
                "role": msg.role,
                "content": self._interpolate(msg.content, variables, trim),
            }
            for msg in self._raw_messages
        ]

    def compile_to_string(self, variables: dict[str, Any] | None = None) -> str:
        """Compile and join all messages into a single string."""
        return "\n\n".join(m["content"] for m in self.compile(variables))

    def system_prompt(self, variables: dict[str, Any] | None = None) -> str | None:
        """Return just the system message content, compiled."""
        sys_msg = next((m for m in self._raw_messages if m.role == "system"), None)
        if not sys_msg:
            return None
        return self._interpolate(sys_msg.content, variables or {}, trim=True)

    async def test(
        self,
        cases: list[TestCase],
        runner: Callable[[list[dict[str, str]]], Awaitable[str]],
        scorer: Callable[[str, str], Awaitable[float]] | None = None,
    ) -> list[TestResult]:
        """
        Run regression tests against this prompt using your LLM runner.

        Example:
            async def runner(messages):
                response = await openai.chat.completions.create(
                    model="gpt-4o", messages=messages
                )
                return response.choices[0].message.content

            results = await prompt.test(test_cases, runner=runner)
            failed = [r for r in results if not r.passed]
        """
        results: list[TestResult] = []
        _scorer = scorer or _default_similarity

        for case in cases:
            try:
                messages = self.compile(case.input)
                actual = await runner(messages)
                score = await _scorer(actual, case.expected_output)
                results.append(TestResult(
                    test_case=case.name,
                    passed=score >= case.threshold,
                    score=score,
                    threshold=case.threshold,
                    actual_output=actual,
                ))
            except Exception as e:
                results.append(TestResult(
                    test_case=case.name,
                    passed=False,
                    score=0.0,
                    threshold=case.threshold,
                    error=str(e),
                ))

        return results

    def __repr__(self) -> str:
        return f"Prompt({self.metadata.slug}@{self.metadata.version}, env={self.metadata.env})"

    # ── Private ──────────────────────────────────

    def _interpolate(self, template: str, variables: dict[str, Any], trim: bool) -> str:
        def replacer(match: re.Match) -> str:  # type: ignore[type-arg]
            key = match.group(1).strip()
            if key in variables:
                return str(variables[key])
            return match.group(0)  # leave unknown placeholders as-is

        result = re.sub(r"\{\{\s*([\w.]+)\s*\}\}", replacer, template)
        return result.strip() if trim else result


async def _default_similarity(a: str, b: str) -> float:
    """
    Jaccard similarity on word tokens.
    Works offline — no embedding API needed.
    For production, replace with an embeddings-based scorer.
    """
    def tokenise(s: str) -> set[str]:
        return set(re.sub(r"[^a-z0-9\s]", "", s.lower()).split())

    set_a = tokenise(a)
    set_b = tokenise(b)

    if not set_a and not set_b:
        return 1.0
    if not set_a or not set_b:
        return 0.0

    intersection = len(set_a & set_b)
    union = len(set_a | set_b)
    return intersection / union

File: packages/sdk-ts/src/prompt.ts
// ─────────────────────────────────────────────
//  PromptOps SDK — Prompt
//  The object returned from client.get()
// ─────────────────────────────────────────────

import type {
  RawPrompt,
  PromptMessage,
  PromptMetadata,
  CompileOptions,
  TestCase,
  TestResult,
  RunTestsOptions,
} from './types'
import { PromptOpsCompileError } from './errors'

export class Prompt {
  readonly metadata: PromptMetadata
  private readonly rawMessages: RawPrompt['messages']

  constructor(raw: RawPrompt) {
    this.metadata = raw.metadata
    this.rawMessages = raw.messages
  }

  /**
   * Compile the prompt template by interpolating variables.
   * Replaces {{variable}} placeholders with the values you provide.
   *
   * @example
   * const messages = prompt.compile({ email: emailBody, tone: 'formal' })
   * const result = await openai.chat.completions.create({ model: 'gpt-4o', messages })
   */
  compile(variables: Record<string, string | number | boolean> = {}, options: CompileOptions = {}): PromptMessage[] {
    // Check all required variables are supplied
    const missing = this.metadata.variables.filter(v => !(v in variables))
    if (missing.length > 0) {
      throw new PromptOpsCompileError(
        `Missing required variables for prompt "${this.metadata.slug}": ${missing.join(', ')}`
      )
    }

    return this.rawMessages.map(msg => ({
      role: msg.role,
      content: this.interpolate(msg.content, variables, options.trim ?? true),
    }))
  }

  /**
   * Like compile(), but returns a single string (system prompt merged into first user message).
   * Useful for providers that don't support the messages format.
   */
  compileToString(variables: Record<string, string | number | boolean> = {}): string {
    return this.compile(variables)
      .map(m => m.content)
      .join('\n\n')
  }

  /**
   * Returns only the system message content (compiled).
   */
  systemPrompt(variables: Record<string, string | number | boolean> = {}): string | undefined {
    const sys = this.rawMessages.find(m => m.role === 'system')
    if (!sys) return undefined
    return this.interpolate(sys.content, variables, true)
  }

  /**
   * Run a suite of test cases against this prompt using your LLM runner.
   *
   * @example
   * const results = await prompt.test(testCases, {
   *   runner: async (messages) => {
   *     const res = await openai.chat.completions.create({ model: 'gpt-4o', messages })
   *     return res.choices[0].message.content ?? ''
   *   }
   * })
   */
  async test(cases: TestCase[], options: RunTestsOptions): Promise<TestResult[]> {
    const results: TestResult[] = []

    for (const tc of cases) {
      try {
        const messages = this.compile(tc.input)
        const actual = await options.runner(messages)
        const scorer = options.scorer ?? defaultCosineSimilarity
        const score = await scorer(actual, tc.expectedOutput)
        const threshold = tc.threshold ?? 0.85

        results.push({
          testCase: tc.name,
          passed: score >= threshold,
          score,
          threshold,
          actualOutput: actual,
        })
      } catch (err) {
        results.push({
          testCase: tc.name,
          passed: false,
          score: 0,
          threshold: tc.threshold ?? 0.85,
          error: err instanceof Error ? err.message : String(err),
        })
      }
    }

    return results
  }

  /**
   * Returns a human-readable summary of this prompt version.
   */
  toString(): string {
    return `Prompt(${this.metadata.slug}@${this.metadata.version}, env=${this.metadata.env})`
  }

  /**
   * Raw JSON representation — useful for logging/debugging.
   */
  toJSON() {
    return {
      metadata: this.metadata,
      messages: this.rawMessages,
    }
  }

  // ── Private ──────────────────────────────────

  private interpolate(
    template: string,
    variables: Record<string, string | number | boolean>,
    trim: boolean
  ): string {
    const result = template.replace(/\{\{(\s*[\w.]+\s*)\}\}/g, (match, key: string) => {
      const trimmedKey = key.trim()
      if (trimmedKey in variables) {
        return String(variables[trimmedKey])
      }
      // Leave unknown variables as-is (they may be optional or for nested templates)
      return match
    })
    return trim ? result.trim() : result
  }
}

// ── Default similarity scorer ─────────────────
// Simple token-overlap Jaccard similarity.
// In production this would call an embeddings API,
// but this works well offline for CI test suites.

function defaultCosineSimilarity(a: string, b: string): Promise<number> {
  const tokenise = (s: string) =>
    new Set(
      s
        .toLowerCase()
        .replace(/[^a-z0-9\s]/g, '')
        .split(/\s+/)
        .filter(Boolean)
    )

  const setA = tokenise(a)
  const setB = tokenise(b)

  if (setA.size === 0 && setB.size === 0) return Promise.resolve(1)
  if (setA.size === 0 || setB.size === 0) return Promise.resolve(0)

  let intersection = 0
  setA.forEach(token => { if (setB.has(token)) intersection++ })

  const union = setA.size + setB.size - intersection
  return Promise.resolve(intersection / union)
}